src/fromager/models.py

```diff
@@ -0,0 +1,244 @@
+            rocm:
+                pre_built: True
+    """
+
+    model_config = MODEL_CONFIG
+
+    name: Package
+    """Canonicalized package name"""
+
+    build_dir: BuildDirectory | None = None
+    """sub-directory with setup.py or pyproject.toml"""
+
+    changelog: Changelog = Field(default_factory=dict)
+    """Changelog entries"""
+
+    env: EnvVars = Field(default_factory=dict)
+    """Common env var for all variants"""
+
+    download_source: DownloadSource = Field(default_factory=DownloadSource)
+    """Alternative source download settings"""
+
+    resolver_dist: ResolverDist = Field(default_factory=ResolverDist)
+    """Resolve distribution version"""
+
+    variants: dict[Variant, PackageVariant] = Field(default_factory=dict)
+    """Variant configuration"""
+
+    @field_validator("download_source", "resolver_dist", "variants", mode="before")
+    @classmethod
+    def before_none_dict(
+        cls, v: _DictStrAny | None, info: ValidationInfo
+    ) -> _DictStrAny:
+        if v is None:
+            v = {}
+        return v
+
+    @classmethod
+    def from_string(cls, package: str | Package, raw_yaml: str) -> "PackageSettings":
+        """Load from raw yaml string"""
+        package = Package(canonicalize_name(package, validate=True))
+        parsed: typing.Any = yaml.safe_load(raw_yaml)
+        if parsed is None:
+            # empty file
+            parsed = {}
+        elif not isinstance(parsed, dict):
+            raise TypeError(f"{package}: invalid yaml, not a dict: {parsed}")
+        return cls(name=package, **parsed)
+
+    @classmethod
+    def from_file(cls, filename: pathlib.Path) -> "PackageSettings":
+        """Load from file
+
+        Raises :exc:`FileNotFound` when the file is not found.
+        The package name is taken from the stem of the file name.
+        """
+        filename = filename.absolute()
+        raw_yaml = filename.read_text(encoding="utf-8")
+        return cls.from_string(filename.stem, raw_yaml)
+
+    @property
+    def override_module_name(self) -> str:
+        """Override module name from package name"""
+        return self.name.replace("-", "_")
+
+    def get_extra_environ(
+        self, variant: Variant, template_env: dict[str, str]
+    ) -> dict[str, str]:
+        """Get extra environment variables for a variant"""
+        # chain entries so variant entries can reference general entries
+        entries = list(self.env.items())
+        pv = self.variants.get(variant)
+        if pv is not None:
+            entries.extend(pv.env.items())
+
+        extra_environ: dict[str, str] = {}
+        for key, value in entries:
+            if isinstance(value, bool):
+                value = "1" if value else "0"
+            elif isinstance(value, int | float):
+                value = str(value)
+            value = value.strip()
+            if "$(" in value:
+                raise ValueError(f"'{value}': subshell '$()' is not supported.")
```

Comment:
In another PR, we should move this validation to when the file is read. Otherwise it might not be triggered for quite a while into a build process.

Reply:
The input is now validated when the config file is loaded.